pkg/spdk/engine.go

```diff
@@ -1128,40 +1131,118 @@ func (e *Engine) ReplicaShallowCopy(dstReplicaName, dstReplicaAddress string) (e
+	if err != nil {
+		return err
+	}
+	dstReplicaServiceCli, err := GetServiceClient(dstReplicaAddress)
+	if err != nil {
+		return err
+	}
+
+	var engineErr error
+	defer func() {
+		if engineErr != nil {
+			if e.State != types.InstanceStateError {
+				e.State = types.InstanceStateError
+				updateRequired = true
 			}
-			if err = srcReplicaServiceCli.ReplicaRebuildingSrcDetach(srcReplicaName, dstReplicaName); err != nil {
-				return err
+			e.ErrorMsg = err.Error()
+			if err == nil {
+				err = engineErr
 			}
-			if err = dstReplicaServiceCli.ReplicaRebuildingDstSnapshotRevert(dstReplicaName, snapshotName); err != nil {
-				return err
+		} else {
+			if e.State != types.InstanceStateError {
+				e.ErrorMsg = ""
+			}
+		}
+
+		// Blindly mark the rebuilding replica as mode ERR now.
+		if err != nil {
+			// Blindly send rebuilding finish for src replica.
+			if srcReplicaErr := srcReplicaServiceCli.ReplicaRebuildingSrcFinish(srcReplicaName, dstReplicaName); srcReplicaErr != nil {
+				e.log.WithError(srcReplicaErr).Errorf("Failed to finish rebuilding for src replica %s with address %s after rebuilding failure, will do nothing", srcReplicaName, srcReplicaAddress)
+			}
+			if e.ReplicaModeMap[dstReplicaName] != types.ModeERR {
+				e.ReplicaModeMap[dstReplicaName] = types.ModeERR
+				e.log.WithError(err).Errorf("Failed to finish rebuilding dst replica %s with address %s from src replica %s with address %s, will mark the rebuilding replica mode as ERR", dstReplicaName, dstReplicaAddress, srcReplicaName, srcReplicaAddress)
+				updateRequired = true
 			}
-			break
 		}
+	}()
+
+	// Pause the IO again by suspending the NVMe initiator
+	if e.Frontend == types.FrontendSPDKTCPBlockdev {
+		if err = e.initiator.Suspend(false, true); err != nil {
+			engineErr = err
+			return errors.Wrapf(err, "failed to suspend NVMe initiator")
+		}
+		defer func() {
+			if engineErr = e.initiator.Resume(); engineErr != nil {
+				return
+			}
+		}()
 	}
 
-	if err = srcReplicaServiceCli.ReplicaRebuildingSrcFinish(srcReplicaName, dstReplicaName); err != nil {
+	// The destination replica will change the parent of the head to the newly rebuilt snapshot chain and detach the external snapshot.
+	// Besides, it should clean up the attached rebuilding lvol if exists.
+	if err = dstReplicaServiceCli.ReplicaRebuildingDstFinish(dstReplicaName); err != nil {
 		return err
 	}
-	// TODO: Connect the previously found latest snapshot lvol with the head lvol chain
-	// TODO: For online rebuilding, the rebuilding lvol must be unexposed
-	if err = dstReplicaServiceCli.ReplicaRebuildingDstFinish(dstReplicaName, e.IP == dstReplicaIP); err != nil {
+	// The source replica will stop exposing the snapshot and wipe the rebuilding info.
+	if err = srcReplicaServiceCli.ReplicaRebuildingSrcFinish(srcReplicaName, dstReplicaName); err != nil {
+		// TODO: Should we mark this healthy replica as error?
 		return err
 	}
 
+	e.ReplicaModeMap[dstReplicaName] = types.ModeRW
```

Comment:
As discussed with @shuo-wu, the mode cannot take effect on the base bdev of the RAID.
We need to improve it and ensure that the IO mode of a base bdev is consistent with the replica mode map in the future.
cc @DamiaSan

Reply:
Let's add TODO and explanation comments for this first.